mediaapi/routing/preview_url.go

```diff
@@ -0,0 +1,400 @@
+	absBasePath config.Path,
+	db storage.Database,
+) *util.JSONResponse {
+	finalPath, duplicate, err := fileutils.MoveFileWithHashCheck(tmpDir, r.MediaMetadata, absBasePath, r.Logger)
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to move file.")
+		return &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.Unknown("Failed to upload"),
+		}
+	}
+	if duplicate {
+		r.Logger.WithField("dst", finalPath).Info("File was stored previously - discarding duplicate")
+	}
+
+	if err = db.StoreMediaMetadata(ctx, r.MediaMetadata); err != nil {
+		r.Logger.WithError(err).Warn("Failed to store metadata")
+		// If the file is a duplicate (has the same hash as an existing file) then
+		// there is valid metadata in the database for that file. As such we only
+		// remove the file if it is not a duplicate.
+		if !duplicate {
+			fileutils.RemoveDir(types.Path(path.Dir(string(finalPath))), r.Logger)
+		}
+		return &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.Unknown("Failed to upload"),
+		}
+	}
+
+	return nil
+}
+
+// parseContent returns the data stored in content attributes
+func parseContent(node *html.Node) string {
+	// iterating the attributes of the tag to get content
+	for _, attr := range node.Attr {
+		if attr.Key == "content" {
+			content := attr.Val
+			return content
+		}
+	}
+	return ""
+}
+
+// prepareResponse prepares the response to be returned to the client
+func (mediaInfo *mediaInfo) prepareResponse(cfg *config.MediaAPI) (*PreviewUrlResponse, *util.JSONResponse) {
+
+	// Reading the file in which the response body was stored and parsing the html.
+	pathToFile, err := fileutils.GetPathFromBase64Hash(mediaInfo.MediaMetadata.Base64Hash, cfg.AbsBasePath)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusInternalServerError,
+			JSON: jsonerror.Unknown("Couldn't get the path to the stored response" + err.Error()),
+		}
+	}
+
+	fileString, err := ioutil.ReadFile(pathToFile)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusInternalServerError,
+			JSON: jsonerror.Unknown("Couldn't read the stored response body" + err.Error()),
+		}
+	}
+
+	file := strings.NewReader(string(fileString))
+
+	tree, err := html.Parse(file)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusInternalServerError,
+			JSON: jsonerror.Unknown("Couldn't get the path to the stored file " + err.Error()),
+		}
+	}
+
+	// map for storing the content values
+	m := make(map[string]string)
+
+	// Iterating the *html.Node, looking for og data
+	var f func(*html.Node)
+	f = func(n *html.Node) {
+		//check if meta tag
+		if n.Type == html.ElementNode && n.Data == "meta" {
+			//parse attributes of the tag
+			for _, a := range n.Attr {
+				if a.Key == "property" && strings.HasPrefix(a.Val, "og:") {
+					slice := strings.Split(a.Val, ":")
+					m[slice[1]] = parseContent(n)
+				}
+			}
+		}
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			f(c)
```

Comment:
We probably want to limit how much we're willing to explore on the HTML to avoid pathological cases which consume lots of CPU.